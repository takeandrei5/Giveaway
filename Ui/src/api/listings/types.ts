import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
