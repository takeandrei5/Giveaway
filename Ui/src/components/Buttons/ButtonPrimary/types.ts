import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
