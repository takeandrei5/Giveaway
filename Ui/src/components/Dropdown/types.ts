export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
