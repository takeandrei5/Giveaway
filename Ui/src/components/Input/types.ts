export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
