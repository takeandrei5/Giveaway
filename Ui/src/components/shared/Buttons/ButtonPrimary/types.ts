import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
