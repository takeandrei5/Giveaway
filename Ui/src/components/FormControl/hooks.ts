import { FieldMetaProps } from 'formik';
import { useEffect, useState } from 'react';

import { useCheckFormIsInvalidProps } from './types';

const useCheckFormIsInvalid = (meta: FieldMetaProps<unknown>): useCheckFormIsInvalidProps => {
	const [isInvalid, setIsInvalid] = useState(false);

	useEffect(() => {
		setIsInvalid(meta.touched && !!meta.error);
	}, [meta.touched, meta.error]);

	return { isInvalid };
};

export { useCheckFormIsInvalid };
