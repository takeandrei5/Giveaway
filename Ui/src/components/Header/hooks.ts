import { NextRouter, useRouter } from 'next/router';
import { useCallback } from 'react';

const useLogin = () => {
	const router: NextRouter = useRouter();

	const handleSignIn: () => Promise<boolean> = useCallback(
		async () => await router.replace('/api/auth/login'),
		[router]
	);
	const handleSignInWithReturnTo: (returnTo: string) => Promise<boolean> = useCallback(
		async (returnTo: string) => await router.replace(`/api/auth/login?returnTo=${returnTo}`),
		[router]
	);
	const handleSignOut: () => Promise<boolean> = useCallback(() => router.replace('/api/auth/logout'), [router]);

	return { handleSignIn, handleSignInWithReturnTo, handleSignOut };
};

export { useLogin };
