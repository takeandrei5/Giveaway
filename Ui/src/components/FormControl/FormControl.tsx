import { FormControl as FormControlChakra, FormErrorMessage, FormLabel, useColorModeValue } from '@chakra-ui/react';
import { useField } from 'formik';
import React from 'react';

import { useCheckFormIsInvalid } from './hooks';
import { FormControlProps } from './types';

const FormControl = ({ children, id, label, name }: FormControlProps): JSX.Element => {
	const [_, meta] = useField<unknown>(name);
	const { isInvalid } = useCheckFormIsInvalid(meta);

	const darkOrLightColor: 'dark' | 'light' = useColorModeValue('dark', 'light');

	return (
		<FormControlChakra isInvalid={isInvalid} isRequired width='100%'>
			<FormLabel htmlFor={id} margin='0' color={darkOrLightColor}>
				{label}
			</FormLabel>
			{React.cloneElement(children, { isInvalid })}
			{isInvalid && (
				<FormErrorMessage data-testid='form-control-error-message' marginTop='0.125rem' color='danger'>
					{meta.error}
				</FormErrorMessage>
			)}
		</FormControlChakra>
	);
};

export default FormControl;
