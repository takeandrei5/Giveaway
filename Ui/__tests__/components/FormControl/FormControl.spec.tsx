import { FormControl } from '@components';
import { useCheckFormIsInvalid } from '@components/FormControl/hooks';
import { FormControlProps } from '@components/FormControl/types';
import { render } from '@testing-library/react';
import { FieldMetaProps, useField } from 'formik';
import React from 'react';

jest.mock('formik', () => ({
	...jest.requireActual('formik'),
	useField: jest.fn(),
}));

jest.mock('@components/FormControl/hooks', () => ({
	useCheckFormIsInvalid: jest.fn((meta: FieldMetaProps<unknown>) => ({})),
}));

describe('FormControl', () => {
	it('should match snapshot', () => {
		// Arrange
		const props: FormControlProps = {
			id: 'test-id',
			label: 'test-label',
			name: 'test-name',
			children: <></>,
		};

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				onBlur: jest.fn(),
				onChange: jest.fn(),
				value: 'test-name',
			},
			{
				value: 'test-name',
				error: '',
				touched: false,
				initialValue: 'test-name',
				initialTouched: false,
				initialError: '',
			},
		]);

		// Act
		const component = render(<FormControl {...props} />);

		// Assert
		expect(component).toMatchSnapshot();
	});

	it('should render error message if invalid', () => {
		// Arrange
		const props: FormControlProps = {
			id: 'test-id',
			label: 'test-label',
			name: 'test-name',
			children: <></>,
		};

		const meta: FieldMetaProps<unknown> = {
			value: 'test-name',
			error: 'error',
			touched: false,
			initialValue: 'test-name',
			initialTouched: false,
			initialError: '',
		};

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				onBlur: jest.fn(),
				onChange: jest.fn(),
				value: 'test-name',
			},
			meta,
		]);

		(useCheckFormIsInvalid as unknown as jest.Mock).mockImplementation((meta: FieldMetaProps<unknown>) => ({
			isInvalid: true,
		}));

		// Act
		const { getByTestId } = render(<FormControl {...props} />);

		// Assert
		expect(getByTestId('form-control-error-message')).toBeTruthy();
	});

	it('should render children with the correct isInvalid prop', () => {
		// Arrange
		const TestComponent = ({ isInvalid = false }: any) => {
			return <div data-testid='children-test-id' id={`isInvalid-${isInvalid ? true : false}`} />;
		};

		const props: FormControlProps = {
			id: 'test-id',
			label: 'test-label',
			name: 'test-name',
			children: <TestComponent />,
		};

		const meta: FieldMetaProps<unknown> = {
			value: 'test-name',
			error: 'error',
			touched: false,
			initialValue: 'test-name',
			initialTouched: false,
			initialError: '',
		};

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				onBlur: jest.fn(),
				onChange: jest.fn(),
				value: 'test-name',
			},
			meta,
		]);

		(useCheckFormIsInvalid as unknown as jest.Mock).mockImplementation((meta: FieldMetaProps<unknown>) => ({
			isInvalid: true,
		}));

		// Act
		const { getByTestId } = render(<FormControl {...props} />);

		// Assert
		expect(getByTestId('children-test-id').id).toBe('isInvalid-true');
	});
});
