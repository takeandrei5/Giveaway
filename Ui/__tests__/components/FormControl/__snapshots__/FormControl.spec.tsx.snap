// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl should match snapshot 1`] = `
<div>
  <div
    class="chakra-form-control css-8atqhb"
    role="group"
  >
    <label
      class="chakra-form__label css-xgxc5v"
      for="test-id"
      id="field-1-label"
    >
      test-label
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
  </div>
</div>
`;
